### PR TITLE
Unset deprecated flags from 1.11

### LINKED ISSUE #425: APIserver admission controller convert to `--enable-admission-plugins` and `--disable-admission-plugins`
These flag `admission-controllers` is deprecated and for clusters >= 1.11 we should use:

 `--enable-admission-plugins` and `--disable-admission-plugins`

The order of admisson controlller set with those flags are determined by the API server automatically

/kind technical-debt

### LINKED ISSUE #426: APIserver remove `--insecure-port` and `--etc-quorum-read` >= 1.11
These flags are deprecated and for clusters >= 1.11 we should not set them any more.

/kind technical-debt

### PR DESCRIPTION
**Pull request by JoshVanL**
**What this PR does / why we need it**:
Unset depreciated flags for Kubernetes 1.11

fixes #425 
fixes #426 

```release-note
Unset API Server depreciated flags for Kubernetes version >= 1.11
```
/assign @simonswine

### CODE REVIEW

File: puppet/modules/kubernetes/manifests/apiserver.pp [hunk 1]
```diff
@@ -101,7 +106,10 @@
   }
 
   # Do not insecure bind the API server on kubernetes 1.6+
-  $insecure_port = $::kubernetes::_apiserver_insecure_port
+  if !$post_1_11 {
+    $insecure_port = $::kubernetes::_apiserver_insecure_port
+    $etcd_quorum_read = true
+  }
```
Comment (simonswine) [inline puppet/modules/kubernetes/manifests/apiserver.pp]: comment needs updating

File: puppet/modules/kubernetes/manifests/init.pp [hunk 2]
```diff
@@ -52,6 +52,7 @@
   }
 
   # do not insecure bind the apiserver after 1.5
+
```
Comment (simonswine) [inline puppet/modules/kubernetes/manifests/init.pp]: remove me

File: puppet/modules/kubernetes/templates/kube-apiserver.service.erb [hunk 3]
```diff
@@ -78,7 +80,12 @@ ExecStart=<%= scope['kubernetes::_dest_dir'] %>/apiserver \
   --etcd-keyfile=<%= scope['kubernetes::apiserver::etcd_key_file'] %> \
 <% end -%>
   --service-cluster-ip-range=<%= scope['kubernetes::service_ip_range_network'] %>/<%= scope['kubernetes::service_ip_range_mask'] %> \
+<% if @post_1_11 -%>
+  --enable-admission-plugins=<%= @_admission_control.join(',') %> \
+  --disable-admission-plugins=<%= @_disable_admission_control.join(',') %> \
+<%- else -%>
```
Comment (simonswine) [inline puppet/modules/kubernetes/templates/kube-apiserver.service.erb]: This is not really backwards compatible yet, is it?

We we build the old admission_control thing we need to filter out the default enabled ones post 1.11.

Somehow we need to expose the disable admission controller flags

**Comment by simonswine** [comment]
/assign @JoshVanL 
/unassign

Please add a release-note

File: puppet/modules/kubernetes/manifests/apiserver.pp [hunk 4]
```diff
@@ -100,8 +112,11 @@
     $_oidc_signing_algs = []
   }
 
-  # Do not insecure bind the API server on kubernetes 1.6+
-  $insecure_port = $::kubernetes::_apiserver_insecure_port
+  # Do not insecure bind the API server on kubernetes 1.11+
```
Comment (MattiasGees) [inline puppet/modules/kubernetes/manifests/apiserver.pp]: Should be `Do not set insecure_port variable of the API server on kubernetes 1.11+`

File: puppet/modules/kubernetes/manifests/apiserver.pp [hunk 5]
```diff
@@ -100,8 +112,11 @@
...
+  if !$post_1_11 {
+    $insecure_port = $::kubernetes::_apiserver_insecure_port
```
Comment (MattiasGees) [inline puppet/modules/kubernetes/manifests/apiserver.pp]: Add comment `# Do not insecure bind the API server on kubernetes 1.6 to 1.10`

File: puppet/modules/kubernetes/manifests/apiserver.pp [hunk 6]
```diff
@@ -51,6 +52,8 @@
   $_systemd_after = ['network.target'] + $systemd_after
   $_systemd_before = $systemd_before
 
+  $post_1_11 = versioncmp($::kubernetes::version, '1.11.0') >= 0
+  $post_1_10 = versioncmp($::kubernetes::version, '1.10.0') >= 0
```
Comment (MattiasGees) [inline puppet/modules/kubernetes/manifests/apiserver.pp]: Not used anywhere?

File: docs/user-guide.rst [hunk 7]
```diff
@@ -563,6 +563,24 @@ by limiting the access to a list of CIDR blocks. This can be configured on a
...
+Enabled and disabled plugins can be configured as follows:
+
+.. code-block:: yaml
+
+  kubernetes:
+    apiServer:
+      enableAdmissionControllers: ["DefaultStorageClass", "DefaultTolerationSeconds"]
+      disableAdmissionControllers: ["MutatingAdmissionWebhook"]
+
+**Note:** Disabling admission control plugins is only available with Kubernetes
```
[the remark this answers is not in the document]
Comment (JoshVanL) [reply docs/user-guide.rst]: Sure, not sure about having a check if they exist. Will be difficult to keep that list up-to-date imo.

[the remark this answers is not in the document]
Comment (MattiasGees) [reply docs/user-guide.rst]: I think we should do this. I tried setting up a kubernetes cluster with a non-existing admission controller. This caused kubernetes-api to crash are systemd retrying to start it.

Error message I got:

```
Oct 19 13:32:43 ip-10-99-50-202.eu-west-1.compute.internal systemd[1]: Started Kubernetes API Server.
Oct 19 13:32:43 ip-10-99-50-202.eu-west-1.compute.internal systemd[1]: Starting Kubernetes API Server...
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: Flag --insecure-port has been deprecated, This flag will be removed in a future version.
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: Flag --etcd-quorum-read has been deprecated, This flag is deprecated and the ability to switch off quorum read will be removed in a future release.
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: Flag --admission-control has been deprecated, Use --enable-admission-plugins or --disable-admission-plugins instead. Will be removed in a future version.
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283304   27615 flags.go:27] FLAG: --address="127.0.0.1"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283364   27615 flags.go:27] FLAG: --admission-control="[DefaultStorageClass,test]"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283382   27615 flags.go:27] FLAG: --admission-control-config-file=""
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283393   27615 flags.go:27] FLAG: --advertise-address="<nil>"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283401   27615 flags.go:27] FLAG: --allow-privileged="true"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283412   27615 flags.go:27] FLAG: --allow-verification-with-non-compliant-keys="false"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283423   27615 flags.go:27] FLAG: --alsologtostderr="false"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283430   27615 flags.go:27] FLAG: --anonymous-auth="true"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283438   27615 flags.go:27] FLAG: --apiserver-count="1"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283449   27615 flags.go:27] FLAG: --application-metrics-count-limit="100"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283458   27615 flags.go:27] FLAG: --audit-log-batch-buffer-size="10000"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283466   27615 flags.go:27] FLAG: --audit-log-batch-max-size="400"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283473   27615 flags.go:27] FLAG: --audit-log-batch-max-wait="30s"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283484   27615 flags.go:27] FLAG: --audit-log-batch-throttle-burst="15"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283492   27615 flags.go:27] FLAG: --audit-log-batch-throttle-enable="false"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283500   27615 flags.go:27] FLAG: --audit-log-batch-throttle-qps="10"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283512   27615 flags.go:27] FLAG: --audit-log-format="json"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283520   27615 flags.go:27] FLAG: --audit-log-maxage="0"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283528   27615 flags.go:27] FLAG: --audit-log-maxbackup="1"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283536   27615 flags.go:27] FLAG: --audit-log-maxsize="100"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283544   27615 flags.go:27] FLAG: --audit-log-mode="blocking"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283553   27615 flags.go:27] FLAG: --audit-log-path="/var/log/kubernetes/audit.log"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283562   27615 flags.go:27] FLAG: --audit-log-truncate-enabled="false"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283570   27615 flags.go:27] FLAG: --audit-log-truncate-max-batch-size="10485760"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283582   27615 flags.go:27] FLAG: --audit-log-truncate-max-event-size="102400"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283590   27615 flags.go:27] FLAG: --audit-policy-file="/etc/kubernetes/audit-policy.yaml"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283599   27615 flags.go:27] FLAG: --audit-webhook-batch-buffer-size="10000"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283606   27615 flags.go:27] FLAG: --audit-webhook-batch-initial-backoff="10s"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283615   27615 flags.go:27] FLAG: --audit-webhook-batch-max-size="400"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283624   27615 flags.go:27] FLAG: --audit-webhook-batch-max-wait="30s"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283633   27615 flags.go:27] FLAG: --audit-webhook-batch-throttle-burst="15"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283641   27615 flags.go:27] FLAG: --audit-webhook-batch-throttle-enable="true"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283649   27615 flags.go:27] FLAG: --audit-webhook-batch-throttle-qps="10"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283658   27615 flags.go:27] FLAG: --audit-webhook-config-file=""
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283667   27615 flags.go:27] FLAG: --audit-webhook-initial-backoff="10s"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283676   27615 flags.go:27] FLAG: --audit-webhook-mode="batch"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283684   27615 flags.go:27] FLAG: --audit-webhook-truncate-enabled="false"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283692   27615 flags.go:27] FLAG: --audit-webhook-truncate-max-batch-size="10485760"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283700   27615 flags.go:27] FLAG: --audit-webhook-truncate-max-event-size="102400"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283709   27615 flags.go:27] FLAG: --authentication-token-webhook-cache-ttl="2m0s"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283717   27615 flags.go:27] FLAG: --authentication-token-webhook-config-file=""
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283725   27615 flags.go:27] FLAG: --authorization-mode="Node,RBAC"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283733   27615 flags.go:27] FLAG: --authorization-policy-file=""
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283760   27615 flags.go:27] FLAG: --authorization-rbac-super-user=""
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283769   27615 flags.go:27] FLAG: --authorization-webhook-cache-authorized-ttl="5m0s"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283778   27615 flags.go:27] FLAG: --authorization-webhook-cache-unauthorized-ttl="30s"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283786   27615 flags.go:27] FLAG: --authorization-webhook-config-file=""
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283795   27615 flags.go:27] FLAG: --azure-container-registry-config=""
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283803   27615 flags.go:27] FLAG: --basic-auth-file=""
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283811   27615 flags.go:27] FLAG: --bind-address="0.0.0.0"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283820   27615 flags.go:27] FLAG: --boot-id-file="/proc/sys/kernel/random/boot_id"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283829   27615 flags.go:27] FLAG: --cert-dir="/var/run/kubernetes"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283837   27615 flags.go:27] FLAG: --client-ca-file="/etc/kubernetes/ssl/kube-apiserver-ca.pem"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283846   27615 flags.go:27] FLAG: --cloud-config=""
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283853   27615 flags.go:27] FLAG: --cloud-provider="aws"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283861   27615 flags.go:27] FLAG: --cloud-provider-gce-lb-src-cidrs="130.211.0.0/22,209.85.152.0/22,209.85.204.0/22,35.191.0.0/16"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283876   27615 flags.go:27] FLAG: --container-hints="/etc/cadvisor/container_hints.json"
Oct 19 13:32:44 ip-10-99-50-202.eu-west-1.compute.internal apiserver[27615]: I1019 13:32:44.283885   27615 flags.go:27] FLAG: --containerd="unix:///var/run/containerd.sock"
```

File: docs/user-guide.rst [hunk 8]
```diff
@@ -563,6 +563,24 @@ by limiting the access to a list of CIDR blocks. This can be configured on a
...
+API admission control plugins can be enabled and disabled through the Tarmak
+configuration file.
+`Information on admission plugins can be found here <https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/>`_.
+Enabled and disabled plugins can be configured as follows:
+
+.. code-block:: yaml
+
+  kubernetes:
+    apiServer:
+      enableAdmissionControllers: ["DefaultStorageClass", "DefaultTolerationSeconds"]
```
Comment (MattiasGees) [inline docs/user-guide.rst]: I think we should show it as:
```
kubernetes:
  apiServer:
    enableAdmissionControllers:
      - DefaultStorageClass
      - DefaultTolerationSeconds
```

File: docs/user-guide.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (JoshVanL) [reply docs/user-guide.rst]: Not sure that tarmak should be responsible for checking what is and isn't available imo

[the remark this answers is not in the document]
Comment (MattiasGees) [reply docs/user-guide.rst]: True, maybe we should add a warning in the docs and also points to the kubernetes docs for the available admission plugins?

**Comment by JoshVanL** [comment]
@MattiasGees Interesting, guess filter doesn't work quite as expected
```
  if $post_1_11 {
    $defaults = ['NamespaceLifecycle','LimitRanger','ServiceAccount','PersistentVolumeLabel','DefaultStorageClass','DefaultTolerationSeconds','MutatingAdmissionWebhook','ValidatingAdmissionWebhook','ResourceQuota','Priority']
    $_admission_control = $unfiltered_admission_control.filter |$value| { !($value in $defaults) }
  } else {
    $_admission_control = $unfiltered_admission_control
  }
```

**Comment by MattiasGees** [comment]
Everything works now, tested with k8s 1.10 and 1.11. Good good behaviour back. When no admission controller is set in `tarmak.yaml`, Tarmak uses the default. 

As discussed in chat @JoshVanL, just some extra docs needed. We have to document that we set some default admission controllers by default. Once the user set one admission controller, Tarmak will only set that one and remove the defaults.

**Comment by simonswine** [comment]
Not entirely sure this should be ready, seems to work for me (@JoshVanL )

/unassign @MattiasGees 
/assign

**Pull request merged**